drivers/samsung/displays/md_series.cr

```diff
@@ -0,0 +1,500 @@
+        self[:input] = INPUTS.new(value).to_s
+        # The input feedback behaviour seems to go a little odd when
+        # screen split is active. Ignore any input forcing when on.
+        unless self[:screen_split]
+          self[:input_stable] = self[:input] == self[:input_target]
+          switch_to(self[:input_target].as_s) unless self[:input_stable]
+        end
+      when COMMAND::Speaker
+        self[:speaker] = SPEAKERMODES.new(value).to_s
+      when COMMAND::Hard_off
+        self[:hard_off] = value == 0
+        self[:power] = false if self[:hard_off]
+      when COMMAND::Screen_split
+        self[:screen_split] = value >= 0
+      when COMMAND::Software_version
+        self[:software_version] = values.join
+      when COMMAND::Serial_number
+        self[:serial_number] = values.join
+      else
+        logger.debug { "Samsung responded with ACK: #{value}" }
+      end
+
+      task.try &.success
+    when RESPONSESTATUS::Nak
+      task.try &.abort("Samsung responded with NAK: #{hex}")
+    else
+      task.try &.abort("Samsung aborted with: #{hex}")
+    end
+  end
+
+  def check_power_state
+    return if self[:power_stable]?
+    if self[:power]? == self[:power_target]?
+      self[:power_stable] = true
+    else
+      power(self[:power_target].as_bool)
+    end
+  end
+
+  enum COMMAND
+    Status           = 0x00
+    Hard_off         = 0x11 # Completely powers off
+    Panel_mute       = 0xF9 # Screen blanking / visual mute
+    Volume           = 0x12
+    Contrast         = 0x24
+    Brightness       = 0x25
+    Sharpness        = 0x26
+    Colour           = 0x27
+    Tint             = 0x28
+    Red_gain         = 0x29
+    Green_gain       = 0x2A
+    Blue_gain        = 0x2B
+    Input            = 0x14
+    Mode             = 0x18
+    Size             = 0x19
+    Pip              = 0x3C # picture in picture
+    Auto_adjust      = 0x3D
+    Wall_mode        = 0x5C # Video wall mode
+    Safety           = 0x5D
+    Wall_on          = 0x84 # Video wall enabled
+    Wall_user        = 0x89 # Video wall user control
+    Speaker          = 0x68
+    Net_standby      = 0xB5 # Keep NIC active in standby
+    Eco_solution     = 0xE6 # Eco options (auto power off)
+    Auto_power       = 0x33
+    Screen_split     = 0xB2 # Tri / quad split (larger panels only)
+    Software_version = 0x0E
+    Serial_number    = 0x0B
+    Time             = 0xA7
+  end
+
+  private def do_send(command : String | Int, data : Int | Array(Int) = [] of UInt8, **options)
+    data = [data] if data.is_a?(Int)
+
+    # # options[:name] = command if data.length > 0 # name unless status request
+    command = COMMAND.parse(command).value if command.is_a?(String)
+
+    data = [command, @id, data.size] + data # Build request
+    data << (data.sum & 0xFF)               # Add checksum
+    data = [0xAA] + data                    # Add header
+
+    # Convert to Bytes
+    data = Slice.new(data.size) { |i| data[i].to_u8 }
```

Comment:
There's a quite a few intermediate arrays being created here. Take a looks at [this method for creating Slices](https://crystal-lang.org/api/0.35.1/Slice.html#new(size:Int,*,read_only=false,&)-class-method).

Reply:
Hey @KimBurgess , I've made this change as I think it may be cleaner than creating a new slice and using a block to do things. Let me know what you think.

I haven't found a way to copy an array or even another Slice into a Slice so I've settled on doing this https://github.com/PlaceOS/drivers/commit/f5fc9e9798f632b4d832686969a3d4a8bddd0d0d#diff-96bf90ab842197ca263274d3fdb8e648R473

Reply:
Nice! I really like that. Makes it super clear with the mapping to those different byte offsets in the protocol.

If you wanted to wrap this up a little further, you can shift the command assembly into the Command Enum itself and do away with do_send altogether. Something like:
```crystal
enum Command
  Power = 0x11
  Foo   = 0x01
  Bar   = 0x02
  Baz   = 0x03
  
  def build(*data)
    build id: 0xFE_u8, data: data
  end
  
  def build(id : UInt8, data : Tuple) : Bytes
    Bytes.new(data.size + 5).tap do |bytes|
      bytes[0] = 0xAA_u8
      bytes[1] = self.to_u8
      bytes[2] = id
      bytes[3] = data.size.to_u8
      data.each_with_index(4) { |b, i| bytes[i] = b.to_u8 }
      bytes[-1] = bytes[1..-2].reduce &.+.to_u8!
    end
  end
end
```
https://play.crystal-lang.org/#/r/9bp6

Reply:
oh wow, that's neat. Thanks @KimBurgess and also for picking up that I wasn't calculating the checksum correctly. Didn't know how to use the reduce method in crystal so I just used the sum but and took the last 2 bytes which I can see now is not equivalent.